### PR TITLE
Switch OpenTSDB telnet writer to use GenericKeyedObjectPool

### PR DESCRIPTION
**Pull request by ivan-gusev**
And eliminate LifecycleExceptions where possible.
Most of the code and tests is borrowed from GraphiteWriter that trust connection handling to GenericKeyedObjectPool. Prior implementation was throwing LifecycleExceptions on innocent things like refused connections.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23issuecomment-107332031%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23issuecomment-107678333%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23discussion_r31403460%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23discussion_r31403569%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23discussion_r31403607%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23discussion_r31460670%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23discussion_r31460672%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23discussion_r31460673%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23issuecomment-107680750%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23issuecomment-107681053%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23issuecomment-107682098%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23issuecomment-107682576%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23issuecomment-107332031%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Thanks%20a%20lo%20for%20the%20help%20%21%20That%20something%20I%20wanted%20to%20clean%20up.%20I%27ll%20have%20a%20closer%20look%20this%20evening.%22%2C%20%22created_at%22%3A%20%222015-06-01T07%3A06%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-06-01T19%3A28%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3947391%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ivan-gusev%22%7D%7D%2C%20%7B%22body%22%3A%20%22Just%20a%20side%20note%20for%20next%20time%3A%20rebasing%20or%20squashing%20should%20probably%20be%20done%20at%20the%20end%20of%20code%20review%2C%20so%20that%20we%20can%20keep%20track%20of%20the%20discussion.%20Here%2C%20we%20have%20lost%20some%20of%20the%20context%20of%20the%20comments.%22%2C%20%22created_at%22%3A%20%222015-06-01T19%3A40%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%27m%20going%20to%20take%20this%20opportunity%20to%20do%20some%20additional%20clean%20up%20of%20the%20tests%20...%20and%20merge%20all%20this%20...%22%2C%20%22created_at%22%3A%20%222015-06-01T19%3A41%3A41Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%2C%20%7B%22body%22%3A%20%22Sorry%20about%20rebasing.%20I%27m%20not%20a%20frequent%20contributor%20on%20github%20so%20might%20be%20missing%20on%20some%20best%20practices.%22%2C%20%22created_at%22%3A%20%222015-06-01T19%3A44%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3947391%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ivan-gusev%22%7D%7D%2C%20%7B%22body%22%3A%20%22No%20problem%21%20And%20each%20project%20has%20its%20own%20rules%20/%20preferences%2C%20so%20it%20is%20hard%20to%20find%20your%20way%20...%22%2C%20%22created_at%22%3A%20%222015-06-01T19%3A45%3A35Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20e98b27e72b952dec651cf853c837387adf2bed47%20jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java%2039%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23discussion_r31403569%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Now%20that%20the%20pool%20is%20injected%20into%20%60OpenTSDBWriter%60%20we%20could%20probably%20remove%20the%20use%20of%20PowerMock.%20While%20it%20is%20not%20directly%20related%20to%20your%20change%2C%20this%20additional%20improvement%20would%20be%20very%20welcomed%20%21%22%2C%20%22created_at%22%3A%20%222015-06-01T07%3A02%3A37Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-06-01T19%3A28%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3947391%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ivan-gusev%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java%3AL2-39%22%7D%2C%20%22Pull%20e98b27e72b952dec651cf853c837387adf2bed47%20jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/TCollectorUDPWriter.java%2027%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23discussion_r31403460%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22That%20should%20probably%20be%20a%20logger%20call%20and%20not%20a%20call%20to%20%60System.out%60.%22%2C%20%22created_at%22%3A%20%222015-06-01T06%3A59%3A31Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-06-01T19%3A28%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3947391%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ivan-gusev%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/TCollectorUDPWriter.java%3AL60-85%22%7D%2C%20%22Pull%20e98b27e72b952dec651cf853c837387adf2bed47%20jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java%20266%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/300%23discussion_r31403607%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Static%20import%20of%20%60Assertions.assertThat%28%29%60%20would%20be%20more%20readable%20%28if%20there%20is%20no%20conflict%20with%20another%20assertion%20framework%29.%22%2C%20%22created_at%22%3A%20%222015-06-01T07%3A03%3A39Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-complete%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-06-01T19%3A28%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3947391%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ivan-gusev%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java%3AL96-185%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/300#issuecomment-107332031'>General Comment</a></b>
- <a href='https://github.com/gehel'></a> Thanks a lo for the help ! That something I wanted to clean up. I'll have a closer look this evening.
- <a href='https://github.com/ivan-gusev'></a> :+1:<a href='#crh-update-none'></a>
- <a href='https://github.com/gehel'></a> Just a side note for next time: rebasing or squashing should probably be done at the end of code review, so that we can keep track of the discussion. Here, we have lost some of the context of the comments.
- <a href='https://github.com/gehel'></a> I'm going to take this opportunity to do some additional clean up of the tests ... and merge all this ...
- <a href='https://github.com/ivan-gusev'></a> Sorry about rebasing. I'm not a frequent contributor on github so might be missing on some best practices.
- <a href='https://github.com/gehel'></a> No problem! And each project has its own rules / preferences, so it is hard to find your way ...
- [x] <a href='#crh-comment-Pull e98b27e72b952dec651cf853c837387adf2bed47 jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/TCollectorUDPWriter.java 27'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/300#discussion_r31403460'>File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/TCollectorUDPWriter.java:L60-85</a></b>
- <a href='https://github.com/gehel'></a> That should probably be a logger call and not a call to `System.out`.
- [x] <a href='#crh-comment-Pull e98b27e72b952dec651cf853c837387adf2bed47 jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java 39'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/300#discussion_r31403569'>File: jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java:L2-39</a></b>
- <a href='https://github.com/gehel'></a> Now that the pool is injected into `OpenTSDBWriter` we could probably remove the use of PowerMock. While it is not directly related to your change, this additional improvement would be very welcomed !
- [x] <a href='#crh-comment-Pull e98b27e72b952dec651cf853c837387adf2bed47 jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java 266'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/300#discussion_r31403607'>File: jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java:L96-185</a></b>
- <a href='https://github.com/gehel'></a> Static import of `Assertions.assertThat()` would be more readable (if there is no conflict with another assertion framework).

<a href='https://www.codereviewhub.com/jmxtrans/jmxtrans/pull/300?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/jmxtrans/jmxtrans/pull/300?approve=1'></a>&nbsp;<a href='https://github.com/jmxtrans/jmxtrans/pull/300'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by gehel** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/jmxtrans/jmxtrans/pull/300'>Refresh</a>.<a href='#crh-update-none'></a>

File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/TCollectorUDPWriter.java [hunk 1]
```diff
@@ -57,6 +60,25 @@ public TCollectorUDPWriter(
...
+	 * @param query   - the query and its results.
+	 * @param results
+	*/
+	@Override
+	public void internalWrite(Server server, Query query, ImmutableList<Result> results) throws Exception {
+		this.startOutput();
+		for (Result result : results) {
+			for (String resultString : resultParser(result)) {
+				if (isDebugEnabled())
+				System.out.println(resultString);
```
Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/TCollectorUDPWriter.java]: That should probably be a logger call and not a call to `System.out`.

File: jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java [hunk 2]
```diff
@@ -2,33 +2,37 @@
...
 import java.io.InputStreamReader;
 import java.net.Socket;
-import java.net.UnknownHostException;
-
-import static org.mockito.Matchers.anyString;
-import static org.mockito.Matchers.contains;
-import static org.mockito.Matchers.eq;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.List;
```
Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java]: Now that the pool is injected into `OpenTSDBWriter` we could probably remove the use of PowerMock. While it is not directly related to your change, this additional improvement would be very welcomed !

File: jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java [hunk 3]
```diff
@@ -92,201 +96,89 @@ public void	setupTest () throws Exception {
...
-		try {
-			this.writer.doWrite(null, this.mockQuery, ImmutableList.of(this.mockResult));
-			Assert.fail("IOException missing");
-		} catch ( IOException ioCaught ) {
-			// Verify.
-			Assert.assertSame(ioExc, ioCaught);
-			Mockito.verify(this.mockLog).error(contains("writing result"), eq(ioExc));
-		}
+		// check that OpenTSDB format is respected
+		Assertions.assertThat(out.toString()).startsWith("put classNameAlias.attributeName ")
```
Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java]: Static import of `Assertions.assertThat()` would be more readable (if there is no conflict with another assertion framework).

**Comment by gehel** [comment]
Thanks a lo for the help ! That something I wanted to clean up. I'll have a closer look this evening.

**Comment by ivan-gusev** [comment]
:+1:<a href='#crh-update-none'></a>

File: jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java (hunk 2 above)
Comment (ivan-gusev) [inline jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java]: :+1:<a href='#crh-update-none'></a>

File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/TCollectorUDPWriter.java (hunk 1 above)
Comment (ivan-gusev) [inline jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/TCollectorUDPWriter.java]: :+1:<a href='#crh-update-none'></a>

File: jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java (hunk 3 above)
Comment (ivan-gusev) [inline jmxtrans-output/jmxtrans-output-core/src/test/java/com/googlecode/jmxtrans/model/output/OpenTSDBWriterTests.java]: :+1:<a href='#crh-update-complete'></a>

**Comment by gehel** [comment]
Just a side note for next time: rebasing or squashing should probably be done at the end of code review, so that we can keep track of the discussion. Here, we have lost some of the context of the comments.

**Comment by gehel** [comment]
I'm going to take this opportunity to do some additional clean up of the tests ... and merge all this ...

**Comment by ivan-gusev** [comment]
Sorry about rebasing. I'm not a frequent contributor on github so might be missing on some best practices.

**Comment by gehel** [comment]
No problem! And each project has its own rules / preferences, so it is hard to find your way ...

**Pull request merged**